source/common/grpc/google_async_client_cache.h

```diff
@@ -0,0 +1,66 @@
+#include <memory>
+
+#include "envoy/grpc/async_client_manager.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/thread_local/thread_local.h"
+
+namespace Envoy {
+namespace Grpc {
+
+// The RawAsyncClient client cache for Google grpc so channel is not created
+// for each request.
+// TODO(fpliu233): Remove when the cleaner and generic solution for gRPC is
+// live. Tracking in #2598 and #13417.
+class AsyncClientCache : public std::enable_shared_from_this<AsyncClientCache> {
+public:
+  AsyncClientCache(AsyncClientManager& async_client_manager, Stats::Scope& scope,
+                   ThreadLocal::SlotAllocator& tls)
+      : async_client_manager_(async_client_manager), scope_(scope), tls_slot_(tls) {}
+  void init(const ::envoy::config::core::v3::GrpcService& grpc_proto_config);
+  const RawAsyncClientSharedPtr getAsyncClient();
+
+private:
+  /**
+   * Per-thread cache.
+   */
+  struct ThreadLocalCache : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalCache(AsyncClientManager& async_client_manager, Stats::Scope& scope,
+                     const ::envoy::config::core::v3::GrpcService& grpc_proto_config) {
+      const AsyncClientFactoryPtr factory =
+          async_client_manager.factoryForGrpcService(grpc_proto_config, scope, true);
+      async_client_ = factory->create();
+    }
+    RawAsyncClientSharedPtr async_client_;
+  };
+
+  AsyncClientManager& async_client_manager_;
+  Stats::Scope& scope_;
+  ThreadLocal::TypedSlot<ThreadLocalCache> tls_slot_;
+};
+
+using AsyncClientCacheSharedPtr = std::shared_ptr<AsyncClientCache>;
+
+class AsyncClientCacheSingleton : public Singleton::Instance {
+public:
+  AsyncClientCacheSingleton() = default;
+  AsyncClientCacheSharedPtr
+  getOrCreateAsyncClientCache(Grpc::AsyncClientManager& async_client_manager, Stats::Scope& scope,
+                              ThreadLocal::SlotAllocator& tls,
+                              const ::envoy::config::core::v3::GrpcService& grpc_proto_config);
+
+private:
+  // The tls slots with client cache stored with key as hash of
+  // envoy::config::core::v3::GrpcService::GoogleGrpc config.
+  // TODO(fpliu233): When new configuration is pushed, the old client cache will not get cleaned up.
+  // Remove when the cleaner and generic solution for gRPC is live. Tracking in #2598 and #13417.
+  absl::flat_hash_map<std::size_t, AsyncClientCacheSharedPtr> async_clients_;
```

Comment:
What guarantees that all calls to getOrCreateAsyncClientCache happen from a single thread?

Do we need a mutex to protect reads/writes to this map?

Reply:
Yeah I missed this. This is a bug. Either this needs a lock or I would probably just move the hash map into the TLS structure and have a single TLS object that owns all cached clients.

Reply:
The reason I don't want the TLS object to own all the cached clients is that I want the AsyncClient to be created at configuration time, not in the callback.

Reply:
I think a mutex to guard it makes sense

Reply:
Oh right this is for Google gRPC with its own thread pool. Right, OK. This is very hard to keep track of. :)

Reply:
I thought over, it's write in https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/ext_authz/config.cc#L58
and read in the callback function created afterwards https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/ext_authz/config.cc#L66, should this pattern cause any race condition?

Reply:
Right, sorry, I'm really getting my wires crossed here. This is all on the main thread. You are fine.